### ISSUE TITLE
Function variables are changed when using numbers and underscores together

### ISSUE THREAD

**Issue by bradcornford** [less/less.js#2462]
When placing numbers before underscores in function variables, these seem to be changed to include a space character between the two.

Take the following less snippet:

```code
@type: 5_large;

.icon-@{type} {
    background-image: ~"url(/img/icon/@{type}.svg)";
}
```

The output is as follows, notice the space between 5 and the underscore:

```code
.icon-5 _large {
  background-image: url(../img/icon/5 _large.svg);
}
```

**Comment by seven-phases-max** [comment]
Confirmed.

**Comment by lukeapage** [comment]
I think it is a difficult one - if keywords began with numbers then we cannot read `5px` as a dimension. I'm not a fan of trying to work out if a variable is
* a string (as in this case - the correct fix is @type: "5_large";0
* a selector
* something meaningfull in css (e.g. dimension, keyword, expression and a million other things)

I think `@type: 5_large;` should give a syntax error - but we don't check we have whitespace after interpretting a node - and it would probably cause issues if we did - but I think we should do that.

What do you think @seven-phases-max

**Comment by seven-phases-max** [comment]
@lukeapage

I think it's #2485, i.e. just missing `_` in the regexp. I.e. since `5foobar` is parsed is a dimension (with `foobar` as unnit) it's OK to parse `5_foobar` as a dimension too, so it's just missing `_` in the re.
---
Though now I can see the same issue applies to values like `5-foo-bar` (which technically could also be parsed as a single dimension value), but that's more difficult case since then it needs to distinguish between `5-foo-bar` and `5-foo-6` thus requiring more complex regexp there (so I did nothing for it in the patch).

**Comment by lukeapage** [comment]
I'm in favour of saying its not a bug and adding a check whitespace function to the parser so it throws an error (perhaps mentioning, should this be a string?)

**Comment by rjgotten** [comment]
You could always just accept only those strings that are defined as actual unit identifiers in the official [CSS3 Values and Units Module](http://www.w3.org/TR/css3-values/), because really: those are the only units that have sane defined behavior in CSS anyway.

**Comment by seven-phases-max** [comment]
I still can see why `5whatever` and `5_whatever` should behave differently. Yes, I realize that formally both are more like of `Anonymous` type rather than of `Dimension` but since `Dimension` can handle it just fine, why especially forbid it? (not counting those additional `5-foo-bar` examples  though).

**Comment by seven-phases-max** [comment]
>(not counting though that the CSS spec. probably never mentions that some new units they hypothetically invent may not start with  _ ).

Hmm, curiously they actually directly allow this (not counting there's no such units currently):
[`dimension-token`](http://www.w3.org/TR/css-syntax-3/#dimension-token-diagram) is `number-token->ident-token` and [`ident-token`](http://www.w3.org/TR/css-syntax-3/#ident-token-diagram) *may* begin with `_`. ;)

**Comment by rjgotten** [comment]
> Hmm, curiously they actually directly allow this

Not *that* curious, actually.

An `ident-token` is, iirc, also used for property declaration names. The leading underscore on property names was a browser hack for targeting old IE versions. Establishing it as valid syntax that browsers should process is a form of *paving the cowpath*. ;)

**Comment by seven-phases-max** [comment]
>An  ident-token  is, iirc, also used for property declaration names. 

Underscore in property, class, keyword etc. etc. names are expected since those *are* identifiers... Units (e.g. `px` or `%`) contrary are never mentioned in the spec. as kind of an `ident` token hence it was quite surprising to see them in the grammar).

**Comment by matthew-dean** [comment]
The spec is actually surprisingly permissive for the purpose of consistency. So yes, a unit can be pretty much any identifier, for the reason that the parsing does not change as units are added. I don't think it's so much about being permissive to IE. CSS gives specific instructions on how things are to be parsed, so I think they considered that in forming the grammar.

**Comment by seven-phases-max** [comment]
So just formally, despite of the original example, the minimal example:
@type: 5_large;
foo {
    bar: @type; // -> 5 _large
}
```
is about to be fixed even if we know for sure there's no such unit... is it?

**Comment by SomMeri** [comment]
Fixed by #2485

**Comment by bradcornford** [comment]
Thanks for this! Seems to fix the issue!
